source/layout.tex

```diff
@@ -65,8 +65,13 @@
 }}}
 \makeatother
 
-\def\pnum
-{\parabullnum{Paras}{0pt}}
+% Register our intent to number the next paragraph. Don't actually number it
+% yet, because we might have a paragraph break before we see its contents (for
+% example, if the paragraph beings with a note or example).
+\def\pnum{%
+\global\def\maybeaddpnum{\global\def\maybeaddpnum{}\parabullnum{Paras}{0pt}}%
+\everypar=\expandafter{\the\everypar\maybeaddpnum}%
```

Comment:
You're never resetting \everypar (which is probably fine, given that a large number of paragraphs should be numbered).

However, that means you can set \everypar once (outside of \pnum) and not reset it to the same value every time.

Otherwise, very nice solution.

Reply:
memoir sets `\everypar` itself at various points that are not necessarily easy to know about in advance. Setting it only once won't work. I'd like to reset it after invoking `\maybeaddpnum`, but it's not clear that there's a reliable way to do that without clobbering memoir's own uses of it.

This approach might end up with a few dozen `\maybeaddpnum` calls per `\par` for subclauses with lots of `\pnum`s (at least `\rSec` resets `\everypar`, so the number is bounded above by the number of `\pnum`s in a subclause). That doesn't seem to cause any real problems so far. If there's a reliable way to reset `\everypar` without clobbering other uses of it, I'd be very happy to use such a thing!

Reply:
Ok, fine-ish, I guess.